controllers/kokumetricsconfig_controller.go

```diff
@@ -635,44 +636,70 @@ func (r *MetricsConfigReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		files, err := dirCfg.Reports.GetFiles()
 		if err == nil && len(files) > 0 {
 			log.Info("packaging files from an old operator version")
-			packageFiles(packager)
+			packageFiles(packager, cr)
+			// after packaging files after an upgrade, truncate the start time so we recollect
+			// all of today's data. This ensures that today's report contains any new report changes.
+			startTime = startTime.Truncate(24 * time.Hour)
 		}
 	}
 
 	// attempt to collect prometheus stats and create reports
 	if err := getPromCollector(r, cr); err != nil {
-		log.Error(err, "failed to get prometheus connection")
+		log.Info("failed to get prometheus connection", "error", err)
+		r.updateStatusAndLogError(ctx, cr)
 		return ctrl.Result{RequeueAfter: time.Minute * 2}, err // give things a break and try again in 2 minutes
 	}
-	originalStartTime, endTime := getTimeRange(ctx, r, cr)
-	startTime := originalStartTime
-	for startTime.Before(endTime) {
-		t := startTime
-		timeRange := promv1.Range{
-			Start: t,
-			End:   t.Add(59*time.Minute + 59*time.Second),
-			Step:  time.Minute,
+
+	for start := startTime; !start.After(endTime); start = start.AddDate(0, 0, 1) {
+		t := start
+		hours := int(endTime.Sub(t).Hours())
+		if hours > HOURS_IN_DAY {
+			hours = HOURS_IN_DAY
 		}
-		collectPromStats(r, cr, dirCfg, timeRange)
-		if startTime.Sub(originalStartTime) == 48*time.Hour {
-			// after collecting 48 hours of data, package the report to compress the files
+		for i := 0; i <= hours; i++ {
+			timeRange := promv1.Range{
+				Start: t,
+				End:   t.Add(59*time.Minute + 59*time.Second),
+				Step:  time.Minute,
+			}
+			if err := collectPromStats(r, cr, dirCfg, timeRange); err != nil {
+				if err == collector.ErrNoData && t.Hour() == 0 && t.Day() != endTime.Day() && r.initialDataCollection {
+					// if there is no data for the first hour of the day, and we are doing the
+					// initial data collection, skip to the next day so we avoid collecting
+					// partial data for a full day. This ensures we are generating a full daily
+					// report upon initial ingest.
+					log.Info("skipping data collection for day", "datetime", timeRange.Start)
+					break
+				}
+			}
+			t = t.Add(1 * time.Hour)
+		}
+
+		if r.initialDataCollection && t.Sub(startTime).Hours() == 96 {
```

Comment:
This might still split into multiple files for the 4 day period? Also anything specific about the 96 mark or just feels right?

Reply:
The idea here was to fit the number of reports into the default max of 30. Assuming a cluster has 90 days of data, this would generate 90*24/96 ~= 23 reports.

If the reports are too big and require splitting, then yea, this logic goes out the window. But it is a best effort to fit a full 90 days into 30 reports.

Reply:
Thinking about this some more, when we split a payload, that is still counted as a single report. Here is the logic for that:
https://github.com/project-koku/koku-metrics-operator/blob/main/packaging/packaging.go#L428-L435
We get the timestamp from the tar.gz and insert into a set, then count the length of the set.

All payloads from a split report use this basename:
https://github.com/project-koku/koku-metrics-operator/blob/main/packaging/packaging.go#L517

which is just `TIMESTAMP-cost-mgmt...`